### PR TITLE
Fix single wildcard match uploads

### PR DESCRIPTION
**Pull request by adreed-msft**
We were initially going to just create a failing test for this and get to it in the refactor. *HOWEVER*, I discovered it was such a trivial fix that I just went right through with it. Best to have it in the June release, anyway.

If you're wondering, no, this change actually doesn't break plan files. It indirectly FIXES this functionality on past versions when resuming.

If you're curious as to _what_ was even happening before, consider the following request:

`PUT https://<account>.blob.core.windows.net/<container>/test/?timeout=901`

The service was inferring our file name as `test`, and thus plopping it under `test/test`. With this change, the request then becomes

`PUT https://<account>.blob.core.windows.net/<container>/test/<file>?timeout=901`

Fixes the following issues: #410, #397

### CODE REVIEW

File: cmd/copyUploadEnumerator.go [hunk 1]
```diff
@@ -30,6 +30,11 @@ func (e *copyUploadEnumerator) enumerate(cca *cookedCopyCmdArgs) error {
 		return fmt.Errorf("cannot find source to upload")
 	}
 
+	// if the user specifies a virtual directory ex: /container_name/extra_path
+	// then we should extra_path as a prefix while uploading
```
Comment (zezha-msft) [inline cmd/copyUploadEnumerator.go]: typo, missing verb

File: cmd/copyUploadEnumerator.go [hunk 2]
```diff
@@ -64,6 +69,9 @@ func (e *copyUploadEnumerator) enumerate(cca *cookedCopyCmdArgs) error {
 			}
 
 			cleanedSource := strings.Replace(listOfFilesAndDirectories[0], common.OS_PATH_SEPARATOR, common.AZCOPY_PATH_SEPARATOR_STRING, -1)
+			if destinationURL.Path[len(destinationURL.Path)-1:] == `/` {
```
Comment (zezha-msft) [inline cmd/copyUploadEnumerator.go]: hmmm I'm a bit confused, why isn't line 53-55 covering this case already?

File: cmd/zt_copy_blob_upload_test.go [hunk 3]
```diff
@@ -28,6 +28,65 @@ import (
 	"strings"
 )
 
+// regular local file->blob upload
+func (s *cmdIntegrationSuite) TestUploadSingleFileToBlobFolder(c *chk.C) {
```
Comment (zezha-msft) [inline cmd/zt_copy_blob_upload_test.go]: Folder => virtual directory

File: cmd/copyUploadEnumerator.go (hunk 1 above)
Comment (adreed-msft) [reply cmd/copyUploadEnumerator.go]: I literally just copied the majority of a working test 👀, fixing the wording

Comment (adreed-msft) [reply cmd/copyUploadEnumerator.go]: oh yikes

File: cmd/copyUploadEnumerator.go (hunk 2 above)
Comment (adreed-msft) [reply cmd/copyUploadEnumerator.go]: urlIsContainerOrShare is... a really odd function to start with. It doesn't even do the job that the rules listed on top of the function even describe in the first place. I'm rewriting it.

**Comment by adreed-msft** [comment]
@zezha-msft Could you hit this up again? I've rewritten the relevant function to well, actually function.

**Comment by JohnRusk** [comment]
Thanks Adele for the description of the change.  It helped me get my head around it.

File: cmd/copyUtil.go [hunk 4]
```diff
@@ -55,26 +54,21 @@ func (copyHandlerUtil) numOfWildcardInURL(url url.URL) int {
...
-	} else if ((!isIPEndpointStyle && numOfSlashes == 1) || (isIPEndpointStyle && numOfSlashes == 2)) && strings.HasSuffix(url.Path, "/") { // this checks if container_name/ was given
-		return true
+	if azblob.NewBlobURLParts(*url).IPEndpointStyleInfo.AccountName == "" {
+		//Typical endpoint style
+		//If there's no slashes after the first, it's a container.
+		//If there's a slash on the end, it's a virtual directory/container.
+		//Otherwise, it's just a blob.
+		return strings.HasSuffix(url.Path, "/") || strings.Count(url.Path[1:], "/") == 0
+	} else {
+		//URL endpoint style: https://IP:port/accountname/container
```
Comment (JohnRusk) [inline cmd/copyUtil.go]: This is an example of IP endpoint style, but its labelled URL style.  I think its just the comment that's wrong. Should say `// IP endpoint style: https://IP:port/accountname/container`. Do you agree?

File: cmd/copyUtil.go [hunk 5]
```diff
@@ -55,26 +54,21 @@ func (copyHandlerUtil) numOfWildcardInURL(url url.URL) int {
...
+		//If there's 2 or less slashes after the first, it's a container.
```
Comment (JohnRusk) [inline cmd/copyUtil.go]: I'm not getting why the constant here is 2 (rather than 1).  Can you add tests to copyUtil_tests.go that construct an actual IP style endpoint, and exercise this branch of the code.  That will (a) give test coverage to this branch and (b) prove whether it should be 1 or 2.

Comment (adreed-msft) [reply cmd/copyUtil.go]: So, IP endpoints don't contain the account name in their host. Instead, they contain the account name in their path.

As such, the path takes 2 slashes or less after the first to reach the container.

`/adeledevtest1/proxysync/` for instance, would be the path to my `proxysync` container on an IP endpoint. In comparison, the URL endpoint would be `adeledevtest1.blob.core.windows.net/proxysync/`.

File: cmd/copyUtil.go (hunk 4 above)
Comment (adreed-msft) [reply cmd/copyUtil.go]: Yeah, my mistake when writing that

File: cmd/copyUtil.go (hunk 5 above)
Comment (adreed-msft) [reply cmd/copyUtil.go]: If the constant were just 1, then `/adeledevtest1/proxysync` would be considered a blob by AzCopy, and thus we would not provide the service a file name.

File: cmd/copyUtil_test.go [hunk 6]
```diff
@@ -29,27 +29,39 @@ type copyUtilTestSuite struct{}
...
-	isContainerOrShare = util.urlIsContainerOrShare(&testUrl)
-	c.Assert(isContainerOrShare, chk.Equals, true)
+	isContainer = util.urlIsContainer(&testUrl)
+	c.Assert(isContainer, chk.Equals, true)
+}
+
+func (s *copyUtilTestSuite) TestIPIsContainerOrBlob(c *chk.C) {
+	util := copyHandlerUtil{}
+
+	testIP := url.URL{Host: "127.0.0.1:8256", Path: "/account/container"}
```
Comment (JohnRusk) [inline cmd/copyUtil_test.go]: Need to also test one of the form /account/container/virtualdir  Why, since without that test we have not tested the boundary condition of where (with how many slashes) the IP-form of the logic transitions from returning true to returning false.

File: cmd/copyUtil.go (hunk 5 above)
Comment (JohnRusk) [reply cmd/copyUtil.go]: So it should have one more slash than a URL endpoint, right?  But the constant is not one greater than what is used in URL endpoints, it is currently _two_ greater.  That doesn't smell right.

Comment (adreed-msft) [reply cmd/copyUtil.go]: Good point.

**Comment by adreed-msft** [comment]
Squashing and merging
